### ISSUE TITLE
How to add a new file format

### ISSUE THREAD

**Issue by bztsrc** [fontforge/fontforge#4485]
Hi,

I want to add a new font format to FontForge, [SSFN](https://gitlab.com/bztsrc/scalable-font2/). It has a detailed [specification](https://gitlab.com/bztsrc/scalable-font2/-/blob/master/docs/sfn_format.md), but what I don't know, how the FontForge internals are.

So before I start, I thought it would be best to ask for some advice and best practices:
- First thing first, is there a good, up-to-date doc somewhere about the internals? Or a tutorial on adding a new font format perphaps?
- I'm thinking of taking the sfd.c as reference, as it should have an example on how to use the internals. Is this considered to be a good practice? Or are there better options?
- There's an ANSI C linsfn library, but I don't want to integrate that. I would rather write it from scratch and use the internal structures of FF instead of converting between the two representations.
- I'm not sure what to do about features in the font files that are not supported in FF. For example SSFN allows mixing quadratic and cubic curves within a single path, and vector and bitmap layers in the same glyph. Suggestions?
- SSFN fonts might be compressed with deflate. Is there a common scheme on how to handle that? (I mean that's pretty common for bdf and pcf fonts too, so I wonder if I could use the same functions/library avoiding code duplication if possible.)

In general I would be grateful for any pointers to docs on how to add a new format to FF.

Thanks,
bzt

**Comment by jtanx** [comment]
> First thing first, is there a good, up-to-date doc somewhere about the internals? Or a tutorial on adding a new font format perphaps?

There is [this](https://fontforge.org/docs/techref/splinefont.html), but it hasn't been kept up to date, so you're mostly reliant on inspecting the source to figure out what's going on. 

>  I'm thinking of taking the sfd.c as reference, as it should have an example on how to use the internals. Is this considered to be a good practice? Or are there better options?

Have a look too at the SFRead* methods, like https://github.com/fontforge/fontforge/blob/e34bece10b2acea281ef07736bd040994de3531e/fontforge/parsettf.c#L6342

> I'm not sure what to do about features in the font files that are not supported in FF. For example SSFN allows mixing quadratic and cubic curves within a single path, and vector and bitmap layers in the same glyph. Suggestions?

The options I see are:
* Add the support into FF' internal representation (tricky), or
* e.g. convert all to quadratic (i.e a lossy conversion), or
* Store the data for passthrough but don't allow editing of it

> SSFN fonts might be compressed with deflate. Is there a common scheme on how to handle that? (I mean that's pretty common for bdf and pcf fonts too, so I wonder if I could use the same functions/library avoiding code duplication if possible.)

zlib is a hard dependency so you can just use it. But if you're looking for common functions that wrap zlib, I don't think you'll find any. Have a look at woff.c for reference.

Good luck...

**Comment by bztsrc** [comment]
Hi,

> There is this, but it hasn't been kept up to date, so you're mostly reliant on inspecting the source to figure out what's going on.

Yes, I've found that, but I was hoping for a more recent version. I'll read the source then.

> Have a look too at the SFRead* methods, like 

Thanks!

> The options I see are:

Hmm, I think I'll convert quadratics to cubic then, that's much simpler than finding an approximation for quadratics. I don't want to extend FF's internals, that's way too much of a task.

There's no way for mixing SplineChars and BDFChars if I'm correct, a glyph is either this or that. I'll have to figure out what to do with this.

> zlib is a hard dependency so you can just use it.

I've thought so, thanks. I'll use zlib functions then.

> But if you're looking for common functions that wrap zlib, I don't think you'll find any.

Yep, I was hoping for a function that uncompresses and then checks the magic bytes to determine which file format handler to use, or something like that.

> Good luck...

Thanks, looks like I've found myself a nice challenge :-) I've got answers to all my questions, so I'm going to close this issue. Thanks again!

Cheers,
bzt
